src/blueprint/upstream.rs

```diff
@@ -1,20 +1,142 @@
-use std::ops::Deref;
+use std::collections::BTreeSet;
 
-use super::TryFoldConfig;
-use crate::config::{ConfigSet, Upstream};
-use crate::try_fold::TryFold;
+use derive_setters::Setters;
+
+use crate::config::{self, Batch};
 use crate::valid::{Valid, ValidationError};
 
-pub fn to_upstream<'a>() -> TryFold<'a, ConfigSet, Upstream, String> {
-    TryFoldConfig::<Upstream>::new(|config_set, up| {
-        let upstream = up.merge_right(config_set.deref().upstream.clone());
-        if let Some(ref base_url) = upstream.base_url {
-            Valid::from(
-                reqwest::Url::parse(base_url).map_err(|e| ValidationError::new(e.to_string())),
-            )
-            .map_to(upstream.clone())
-        } else {
-            Valid::succeed(upstream.clone())
-        }
-    })
+#[derive(PartialEq, Eq, Clone, Debug, schemars::JsonSchema)]
+pub struct Proxy {
+    pub url: String,
+}
+
+#[derive(PartialEq, Eq, Clone, Debug, Setters, schemars::JsonSchema)]
+pub struct Upstream {
+    pub pool_idle_timeout: u64,
+    pub pool_max_idle_per_host: usize,
+    pub keep_alive_interval: u64,
+    pub keep_alive_timeout: u64,
+    pub keep_alive_while_idle: bool,
+    pub proxy: Option<Proxy>,
+    pub connect_timeout: u64,
+    pub timeout: u64,
+    pub tcp_keep_alive: u64,
+    pub user_agent: String,
+    pub allowed_headers: BTreeSet<String>,
+    pub base_url: Option<String>,
+    pub http_cache: bool,
+    pub batch: Option<Batch>,
+    pub http2_only: bool,
+}
+
+impl Upstream {
+    pub fn get_pool_idle_timeout(&self) -> u64 {
+        self.pool_idle_timeout
+    }
+    pub fn get_pool_max_idle_per_host(&self) -> usize {
+        self.pool_max_idle_per_host
+    }
+    pub fn get_keep_alive_interval(&self) -> u64 {
+        self.keep_alive_interval
+    }
+    pub fn get_keep_alive_timeout(&self) -> u64 {
+        self.keep_alive_timeout
+    }
+    pub fn get_keep_alive_while_idle(&self) -> bool {
+        self.keep_alive_while_idle
+    }
+    pub fn get_connect_timeout(&self) -> u64 {
+        self.connect_timeout
+    }
+    pub fn get_timeout(&self) -> u64 {
+        self.timeout
+    }
+    pub fn get_tcp_keep_alive(&self) -> u64 {
+        self.tcp_keep_alive
+    }
+    pub fn get_user_agent(&self) -> String {
+        self.user_agent.clone()
+    }
+    pub fn get_enable_http_cache(&self) -> bool {
+        self.http_cache
+    }
+    pub fn get_allowed_headers(&self) -> BTreeSet<String> {
+        self.allowed_headers.clone()
+    }
+    pub fn get_delay(&self) -> usize {
+        self.batch.clone().map_or(0, |b| b.delay)
+    }
+
+    pub fn get_max_size(&self) -> usize {
+        self.batch.clone().map_or(100, |b| b.max_size)
+    }
+
+    pub fn get_http_2_only(&self) -> bool {
+        self.http2_only
+    }
+}
```

Comment:
I think we can drop these functions. The Upstream type in blueprint is already contains the final values.

Reply:
blueprint::Upstream type is used here
https://github.com/tailcallhq/tailcall/pull/1076/files/e534988ed5abb987ef374f20d68eca02acc21170#diff-365e2bd1a664609e9450841169e0b22edb367e0c5426baa58d147a00e25d6f02R11

We can't access the methods here if we drop these
https://github.com/tailcallhq/tailcall/blob/78f90f6643f858da5a71fe40e1b02847c42d263c/src/cli/http.rs#L44-L46

Reply:
Needed a bit of refactoring. I have made the changes for now.